src/test/java/BaseTest.java

```diff
@@ -2,17 +2,75 @@
 import org.openqa.selenium.WebDriver;
 import org.openqa.selenium.WebElement;
 import org.openqa.selenium.chrome.ChromeDriver;
+import org.openqa.selenium.chrome.ChromeOptions;
+import org.openqa.selenium.edge.EdgeDriver;
 import org.openqa.selenium.firefox.FirefoxDriver;
+import org.openqa.selenium.interactions.Actions;
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.openqa.selenium.remote.RemoteWebDriver;
 import org.openqa.selenium.safari.SafariDriver;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.testng.annotations.AfterMethod;
 import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.BeforeSuite;
+import org.testng.annotations.Parameters;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.time.Duration;
 
 
 public class BaseTest {
+    public static WebDriver driver = null;
+    public static WebDriverWait wait = null;
+    public static String url = null;
+    public static Actions actions = null;
 
     @BeforeSuite
-    static void setupClass() {
+    public void setupClass() {
         WebDriverManager.chromedriver().setup();
     }
-
+    @BeforeMethod
+    @Parameters({"baseUrl"})
+    public void launchBrowser(String baseUrl) throws MalformedURLException {
```

Comment:
Please finish the homework!You got this problem, please correct it. 

![image](https://user-images.githubusercontent.com/104761482/225842647-42ef86fc-f83b-4532-859c-5062cc2e0423.png)

Hint 

(**@Optional** String baseUrl)

![image](https://user-images.githubusercontent.com/104761482/225843207-24f8b282-6d89-4197-ad67-10f7e8d6f10a.png)

Reply:
Thank you for feedback. 
Why it will be necessary to use @Optional? What is the purpose?
Thank you

Reply:
Thanks!

If defined parameter is not found in your testng.xml file, The test method will receive the default value which is specified inside the @Optional annotation.